design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,211 @@
+
+The `sourceRGVersions` map's keys will be strings in the format `<resource>.<group>`, e.g. "horizontalpodautoscalers.autoscaling". The values will be APIGroup structs. The API Group struct can be imported from k8s.io/apimachinery/pkg/apis/meta/v1. Order the APIGroup.Versions slices using a sort function copied from `k8s.io/apimachinery/pkg/version`.
+
+```go
+sort.SliceStable(gvs, func(i, j int) bool {
+    return version.CompareKubeAwareVersionStrings(gvs[i].Version, gvs[j].Version) > 0
+})
+```
+
+### Objective 3: List the API group versions supported by the target cluster
+
+Still within the `chooseAPIVersionsToRestore` method, the target cluster's resource group versions can now be obtained.
+
+```go
+targetRGVersions := ctx.discoveryHelper.APIGroups()
+```
+
+Order the APIGroup.Versions slices using a sort function copied from `k8s.io/apimachinery/pkg/version`.
+
+```go
+sort.SliceStable(gvs, func(i, j int) bool {
+    return version.CompareKubeAwareVersionStrings(gvs[i].Version, gvs[j].Version) > 0
+})
+```
+
+### Objective 4: Get the user-defined version priorities
+
+Still within the `chooseAPIVersionsToRestore` method, the user-defined version priorities can be retrieved. These priorities are expected to be in a config map named `enableapigroupversions` in the `velero` namespace. An example config map is
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: enableapigroupversions
+  namespace: velero
+data:
+  restoreResourcesVersionPriority: | -
+    rockbands.music.example.io=v2beta1,v2beta2
+    orchestras.music.example.io=v2,v3alpha1
+    subscriptions.operators.coreos.com=v2,v1
+```
+
+In the config map, the resources and groups and the user-defined version priorities will be listed in the `data.restoreResourcesVersionPriority` field following the following general format: `<group>.<resource>=<version 1>[, <version n> ...]`.
+
+A map will be created to store the user-defined priority versions. The map's keys will be strings in the format `<resource>.<group>`. The values will be APIGroup structs that will be imported from `k8s.io/apimachinery/pkg/apis/meta/v1`. Within the APIGroup structs will be versions in the order that the user provides in the config map. The PreferredVersion field in APIGroup struct will be left empty.
+
+### Objective 5: Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail
+
+Determining the priority will also be done in the `chooseAPIVersionsToRestore` method. Once a version is chosen, it will be stored in a new map of the form `map[string]ChosenGRVersion` where the key is the `<resource>.<group>` and the values are of the `ChosenGrpVersion` struct type (shown below). The map will be saved to the `restore.Context` object in a field called `chosenGrpVersToRestore`.
+
+```go
+type ChosenGrpVersion struct {
```

Comment:
I think this could be completely spelled out in the type name.

Reply:
No problem. I'll name the type to `ChosenGroupVersion`.